### PR TITLE
Replace execution outcome types with new runtime APIs

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

Contracts currently return an `ExecutionOutcome` type from every entrypoint. This type is used to inform the host what messages should be sent, and which channels the application should subscribe to and unsubscribe from. However, it would be simpler to use the runtime types to inform the host of the actions whenever (and if ever) they are needed.

## Proposal

Change the guest and host `Contract` traits to not return execution outcomes from the methods. Instead, the runtime types are used to send messages and manage channel subscriptions.

The runtime now keeps track of the application's outcome in the `call_stack`, and updates it based on the calls it receives from the application. The outcome is then used when the application finishes executing its entrypoint.

## Test Plan

The unit tests were updated to use the new APIs.

## Release Plan

Changes the SDK and the WIT interface, so:
- Need to bump the major/minor version number in the next release of the crates.
- Need to update the developer manual.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: linera-base/src/data_types.rs [hunk 1]
```diff
@@ -172,6 +172,40 @@ pub struct Resources {
     // TODO(#1533): Allow declaring calls to other applications instead of having to count them here.
 }
 
+/// A message that is scheduled to be sent.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[cfg_attr(with_testing, derive(Eq, PartialEq))]
+pub struct OutgoingMessage<Message> {
```
Comment (afck) [inline linera-base/src/data_types.rs:178]: I wonder whether we should give it a different name than `linera-chain::data_types::OutgoingMessage`.
E.g. this one could be `PostedMessage` (as in: the application posted it, but it's not quite in the chain's outbox yet)?
Or we could call it `SdkOutgoingMessage`, and then in `linera-sdk` import it `as OutgoingMessage`?
Not sure what's best. Can also be changed in a later PR, if we decide to!

File: linera-sdk/src/contract/runtime.rs [hunk 2]
```diff
@@ -172,3 +192,69 @@ where
...
+{
+    message: OutgoingMessage<Message>,
+}
+
+impl<Message> MessageSender<Message>
+where
+    Message: Serialize,
+{
+    /// Creates a new [`MessageSender`] instance to send the `message` to the `destination`.
+    pub fn new(destination: Destination, message: Message) -> Self {
```
Comment (afck) [inline linera-sdk/src/contract/runtime.rs:210]: Should we make this `pub(crate)` to make it even harder to accidentally send a message in the wrong format?

File: linera-sdk/src/contract/runtime.rs [hunk 3]
```diff
@@ -172,3 +192,69 @@ where
...
+                authenticated: false,
+                is_tracked: false,
+                grant: Resources::default(),
+                message,
+            },
+        }
+    }
+
+    /// Marks the message to be tracked, so that the sender is notified if the message is received
+    /// or receives the message back if it is rejected.
```
Comment (afck) [inline linera-sdk/src/contract/runtime.rs:223]: I don't think we currently send a notification if it's received.
It's only about bouncing if rejected.

File: linera-execution/tests/wasm.rs [hunk 4]
```diff
@@ -24,10 +24,10 @@ use test_case::test_case;
...
 ///
 /// To update the bytecode files, run `linera-execution/update_wasm_fixtures.sh`.
-#[cfg_attr(feature = "wasmer", test_case(WasmRuntime::Wasmer, 48_382, ExecutionRuntimeConfig::Synchronous; "wasmer"))]
-#[cfg_attr(feature = "wasmer", test_case(WasmRuntime::WasmerWithSanitizer, 49_042, ExecutionRuntimeConfig::Synchronous; "wasmer_with_sanitizer"))]
-#[cfg_attr(with_wasmtime, test_case(WasmRuntime::Wasmtime, 49_042, ExecutionRuntimeConfig::Synchronous; "wasmtime"))]
-#[cfg_attr(with_wasmtime, test_case(WasmRuntime::WasmtimeWithSanitizer, 49_042, ExecutionRuntimeConfig::Synchronous; "wasmtime_with_sanitizer"))]
+#[cfg_attr(feature = "wasmer", test_case(WasmRuntime::Wasmer, 61_530, ExecutionRuntimeConfig::Synchronous; "wasmer"))]
+#[cfg_attr(feature = "wasmer", test_case(WasmRuntime::WasmerWithSanitizer, 62_038, ExecutionRuntimeConfig::Synchronous; "wasmer_with_sanitizer"))]
+#[cfg_attr(with_wasmtime, test_case(WasmRuntime::Wasmtime, 62_038, ExecutionRuntimeConfig::Synchronous; "wasmtime"))]
+#[cfg_attr(with_wasmtime, test_case(WasmRuntime::WasmtimeWithSanitizer, 62_038, ExecutionRuntimeConfig::Synchronous; "wasmtime_with_sanitizer"))]
```
Comment (afck) [inline linera-execution/tests/wasm.rs:30]: (That's much more than I'd expect…?)

**Review by afck — APPROVED** [review]
:star_struck:

File: linera-base/src/data_types.rs (hunk 1 above)
Comment (jvff) [reply linera-base/src/data_types.rs:178]: How about `SendMessageRequest`? That might be more accurate in this context.

File: linera-sdk/src/contract/runtime.rs (hunk 2 above)
Comment (jvff) [reply linera-sdk/src/contract/runtime.rs]: Definitely, thanks for catching this!

File: linera-sdk/src/contract/runtime.rs (hunk 3 above)
Comment (jvff) [reply linera-sdk/src/contract/runtime.rs]: Fixed, thanks!

File: linera-execution/tests/wasm.rs (hunk 4 above)
Comment (jvff) [reply linera-execution/tests/wasm.rs:30]: Yeah, it was certainly a surprise to me. I need to investigate more why this is happening.

**Pull request merged**